Unable to compile CppAD with VisualStudio2019

Hi all. I'm trying to compile this library on Windows using `VisualStudio2019` + `CMake` .

The project is generated without any error. However, when I compile the code on `VisualStudio` I obtain the following errors
```
1>------ Build started: Project: cppad_lib, Configuration: RelWithDebInfo x64 ------
1>cppad_colpack.cpp
1>C:\Users\gromualdi\robot-code\CppAD\cppad_lib\cppad_colpack.cpp(21,5): error C2491: 'CppAD::local::this_routine_should_never_get_called': definition of dllimport function not allowed
1>json_parser.cpp
1>C:\Users\gromualdi\robot-code\CppAD\cppad_lib\json_parser.cpp(23,44): error C2375: 'CppAD::local::graph::json_parser': redefinition; different linkage
1>C:\Users\gromualdi\robot-code\CppAD\include\cppad/local/graph/json_parser.hpp(45): message : see declaration of 'CppAD::local::graph::json_parser'
1>C:\Users\gromualdi\robot-code\CppAD\cppad_lib\json_parser.cpp(26,1): error C2491: 'CppAD::local::graph::json_parser': definition of dllimport function not allowed
1>json_writer.cpp
1>C:\Users\gromualdi\robot-code\CppAD\cppad_lib\json_writer.cpp(20,44): error C2375: 'CppAD::local::graph::json_writer': redefinition; different linkage
1>C:\Users\gromualdi\robot-code\CppAD\include\cppad/local/graph/json_writer.hpp(44): message : see declaration of 'CppAD::local::graph::json_writer'
1>C:\Users\gromualdi\robot-code\CppAD\cppad_lib\json_writer.cpp(23,1): error C2491: 'CppAD::local::graph::json_writer': definition of dllimport function not allowed
1>Generating Code...
1>Done building project "cppad_lib.vcxproj" -- FAILED.
========== Build: 0 succeeded, 1 failed, 2 up-to-date, 0 skipped ==========
```
The error is given by the usage of the `maco` `CPPAD_LIB_EXPORT` (alias of `__declspec(dllimport)`). Indeed as written in the [Microsoft documentation](https://docs.microsoft.com/en-us/troubleshoot/cpp/c2491-define-data-member-dllimport), `__declspec(dllimport)` keyword cannot be applied to implement a function. The issue can be easily solved by moving the keyword to the declaration of the function. 

For instance this patch solves the problem (in mine configuration). If you want I can open a PR, however the problem cannot be considered solved by this `patch`. Indeed  unexplored `preprocessor if branches` may have the similar problem. 
See for instance: 
https://github.com/coin-or/CppAD/blob/650e5df8b2a9bf9195f487eb832de7d0e3a06f2d/cppad_lib/cppad_colpack.cpp#L55
```patch
diff --git a/cppad_lib/cppad_colpack.cpp b/cppad_lib/cppad_colpack.cpp
index dc3327053..c815275f4 100644
--- a/cppad_lib/cppad_colpack.cpp
+++ b/cppad_lib/cppad_colpack.cpp
@@ -17,7 +17,9 @@ in the Eclipse Public License, Version 2.0 are satisfied:
 
 # if CPPAD_HAS_COLPACK == 0
 namespace CppAD { namespace local {
-    CPPAD_LIB_EXPORT void this_routine_should_never_get_called(void)
+    CPPAD_LIB_EXPORT void this_routine_should_never_get_called(void);
+
+    void this_routine_should_never_get_called(void)
     {   CPPAD_ASSERT_UNKNOWN(false); }
 } }
 # else // CPPAD_HAS_COLPACK
diff --git a/cppad_lib/json_parser.cpp b/cppad_lib/json_parser.cpp
index fac417d9f..ac0910cee 100644
--- a/cppad_lib/json_parser.cpp
+++ b/cppad_lib/json_parser.cpp
@@ -20,7 +20,7 @@ in the Eclipse Public License, Version 2.0 are satisfied:
 // documentation for this routine is in the file below
 # include <cppad/local/graph/json_parser.hpp>
 
-CPPAD_LIB_EXPORT void CppAD::local::graph::json_parser(
+void CppAD::local::graph::json_parser(
     const std::string& json      ,
     cpp_graph&         graph_obj )
 {   using std::string;
diff --git a/cppad_lib/json_writer.cpp b/cppad_lib/json_writer.cpp
index 250345c6e..add4b20a9 100644
--- a/cppad_lib/json_writer.cpp
+++ b/cppad_lib/json_writer.cpp
@@ -17,7 +17,7 @@ in the Eclipse Public License, Version 2.0 are satisfied:
 // documentation for this routine is in the file below
 # include <cppad/local/graph/json_writer.hpp>
 
-CPPAD_LIB_EXPORT void CppAD::local::graph::json_writer(
+void CppAD::local::graph::json_writer(
     std::string&                              json                   ,
     const cpp_graph&                          graph_obj              )
 {   using std::string;
diff --git a/include/cppad/local/graph/json_parser.hpp b/include/cppad/local/graph/json_parser.hpp
index 17bcc7c7e..a28d2c045 100644
--- a/include/cppad/local/graph/json_parser.hpp
+++ b/include/cppad/local/graph/json_parser.hpp
@@ -42,7 +42,7 @@ Upon return it is a $cref cpp_ad_graph$$ representation of this function.
 $head Prototype$$
 $srccode%hpp% */
 namespace CppAD { namespace local { namespace graph {
-    void json_parser(
+    CPPAD_LIB_EXPORT void json_parser(
         const std::string&  json      ,
         cpp_graph&          graph_obj
     );
diff --git a/include/cppad/local/graph/json_writer.hpp b/include/cppad/local/graph/json_writer.hpp
index 36a4cb39f..299c56cc0 100644
--- a/include/cppad/local/graph/json_writer.hpp
+++ b/include/cppad/local/graph/json_writer.hpp
@@ -41,7 +41,7 @@ This is a $code cpp_graph$$ object.
 $head Prototype$$
 $srccode%hpp% */
 namespace CppAD { namespace local { namespace graph {
-    void json_writer(
+    CPPAD_LIB_EXPORT void json_writer(
         std::string&       json        ,
         const cpp_graph&   graph_obj
     );
```

I am having trouble building the tests using the Visual Studio 2019. I am using the CMake gui to create the 2019 solution file,loading the solution into Visual Studio. When I try to build check_example_atomic_three I get the following error:

    tangent.cpp
     C:\msys64\home\bradl\repo\cppad.git\include\cppad/local/ad_tape.hpp(73,1): error C2063: 'CppAD::operator *': not a function
     C:\msys64\home\bradl\repo\cppad.git\include\cppad/core/atomic/three_afun.hpp(216): message : see reference to class template instantiation 'CppAD::local::ADTape<Base>' being compiled
    with
        [
            Base=float
        ]

The corresponding line in ad_tape.hpp is 

    friend AD<Base> CppAD::operator * <Base>
        (const AD<Base> &left, const AD<Base> &right);

--- a/cppad_lib/cppad_colpack.cpp
+++ b/cppad_lib/cppad_colpack.cpp
@@ -17,7 +17,7 @@ in the Eclipse Public License, Version 2.0 are satisfied:
 
 # if CPPAD_HAS_COLPACK == 0
 namespace CppAD { namespace local {
-    CPPAD_LIB_EXPORT void this_routine_should_never_get_called(void)
+    void this_routine_should_never_get_called(void)
     {   CPPAD_ASSERT_UNKNOWN(false); }
 } }
 # else // CPPAD_HAS_COLPACK
@@ -52,7 +52,7 @@ the entry with index (i, adolc_pattern[i][k]) is a non-zero
 in the sparsity pattern for the matrix.
 */
 // ----------------------------------------------------------------------
-CPPAD_LIB_EXPORT void cppad_colpack_general(
+void cppad_colpack_general(
     CppAD::vector<size_t>&               color         ,
     size_t                               m             ,
     size_t                               n             ,
@@ -164,7 +164,7 @@ For i = 0 , ... , m-1, and for k = 1, ... ,adolc_pattern[i][0],
 the entry with index (i, adolc_pattern[i][k]) is
 in the sparsity pattern for the symmetric matrix.
 */
-CPPAD_LIB_EXPORT void cppad_colpack_symmetric(
+void cppad_colpack_symmetric(
     CppAD::vector<size_t>&               color         ,
     size_t                               m             ,
     const CppAD::vector<unsigned int*>&  adolc_pattern )

--- a/cppad_lib/json_parser.cpp
+++ b/cppad_lib/json_parser.cpp
@@ -20,7 +20,7 @@ in the Eclipse Public License, Version 2.0 are satisfied:
 // documentation for this routine is in the file below
 # include <cppad/local/graph/json_parser.hpp>
 
-CPPAD_LIB_EXPORT void CppAD::local::graph::json_parser(
+void CppAD::local::graph::json_parser(
     const std::string& json      ,
     cpp_graph&         graph_obj )
 {   using std::string;

--- a/cppad_lib/json_writer.cpp
+++ b/cppad_lib/json_writer.cpp
@@ -17,7 +17,7 @@ in the Eclipse Public License, Version 2.0 are satisfied:
 // documentation for this routine is in the file below
 # include <cppad/local/graph/json_writer.hpp>
 
-CPPAD_LIB_EXPORT void CppAD::local::graph::json_writer(
+void CppAD::local::graph::json_writer(
     std::string&                              json                   ,
     const cpp_graph&                          graph_obj              )
 {   using std::string;

--- a/include/cppad/local/cppad_colpack.hpp
+++ b/include/cppad/local/cppad_colpack.hpp
@@ -59,7 +59,7 @@ it is not the case that both
 This routine tries to minimize, with respect to the choice of colors,
 the number of colors.
 */
-extern void cppad_colpack_general(
+CPPAD_LIB_EXPORT void cppad_colpack_general(
           CppAD::vector<size_t>&         color         ,
     size_t                               m             ,
     size_t                               n             ,
@@ -91,7 +91,7 @@ The properties of this coloring have not yet been determined; see
 Efficient Computation of Sparse Hessians Using Coloring
 and Automatic Differentiation (pdf/ad/gebemedhin14.pdf)
 */
-extern void cppad_colpack_symmetric(
+CPPAD_LIB_EXPORT void cppad_colpack_symmetric(
           CppAD::vector<size_t>&         color         ,
     size_t                               n             ,
     const CppAD::vector<unsigned int*>&  adolc_pattern

--- a/include/cppad/local/graph/json_parser.hpp
+++ b/include/cppad/local/graph/json_parser.hpp
@@ -42,7 +42,7 @@ Upon return it is a $cref cpp_ad_graph$$ representation of this function.
 $head Prototype$$
 $srccode%hpp% */
 namespace CppAD { namespace local { namespace graph {
-    void json_parser(
+    CPPAD_LIB_EXPORT void json_parser(
         const std::string&  json      ,
         cpp_graph&          graph_obj
     );

--- a/include/cppad/local/graph/json_writer.hpp
+++ b/include/cppad/local/graph/json_writer.hpp
@@ -41,7 +41,7 @@ This is a $code cpp_graph$$ object.
 $head Prototype$$
 $srccode%hpp% */
 namespace CppAD { namespace local { namespace graph {
-    void json_writer(
+    CPPAD_LIB_EXPORT void json_writer(
         std::string&       json        ,
         const cpp_graph&   graph_obj
     );

